File: Array/linchuanxin/Algorithm/day22_数组嵌套/day22_leetcode_arrayNesting565.c
//
// Created by andylin on 2020/6/22.
//

#include <stdio.h>
#include <stdbool.h>
#include <malloc.h>

int arrayNesting(int* nums, int numsSize){
    bool *visited = (bool *)calloc(numsSize, sizeof(bool));
    int res = 0;
    for (int i = 0; i < numsSize; ++i) {
        if(!*(visited+i))
        {
            int start = *(nums+i);
            int count = 0;
            do
            {
                start = *(nums+start);
                count++;
                *(visited+start) = true;
            }
            while(*(nums+i) != start);
            res = (res >= count) ? res : count;
        }
    }
    return res;
}


int main()
{
    int num[] = {5,4,0,3,1,6,2};
    printf("out res is:%d\n",arrayNesting(&num,7));
    return 0;
}

File: Array/linchuanxin/Algorithm/day22_0622_数组嵌套/day22_leetcode_arrayNesting565.java
package day22;

public class day22_leetcode_arrayNesting565 {



    public static int arrayNesting(int[] nums) {

        int res = 0;
        for (int i = 0; i < nums.length; i++) {
            int count = 0;
            int start = nums[i];
            do{
                 start = nums[start];
                 count++;
            }while (start != nums[i]);
            res = Math.max(res,count);
        }
        return res;
    }



    public static int arrayNesting2(int[] nums) {

        int res = 0;
        boolean[] visited = new boolean[nums.length];

        for (int i = 0; i < nums.length; i++) {
            if(!visited[i]){
                int start = nums[i],count = 0;
                do {
                    start = nums[start];
                    count++;
                    visited[start] = true;
                }while (start != nums[i]);
                res = Math.max(res,count);
            }
        }
        return res;

    }

    public static void main(String[] args) {
        int[] nums = {5,4,0,3,1,6,2};
        System.out.println(arrayNesting2(nums));

    }
}
